cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java

```diff
@@ -93,30 +95,43 @@ private void setConfig(EntityId entityId,
       ProfileDataset profileDataset = ProfileDataset.get(context, datasetFramework);
       PreferencesDataset preferencesDataset = PreferencesDataset.get(context, datasetFramework);
 
-      // need to get old property and check if it contains profile information
-      Map<String, String> oldProperties = preferencesDataset.getPreferences(entityId);
-      preferencesDataset.setPreferences(entityId, propertyMap);
-      NamespaceId namespaceId =
-        entityId instanceof NamespacedEntityId ? ((NamespacedEntityId) entityId).getNamespaceId() : NamespaceId.SYSTEM;
+      boolean isInstanceLevel = entityId.getEntityType().equals(EntityType.INSTANCE);
+      NamespaceId namespaceId = isInstanceLevel ?
+        NamespaceId.SYSTEM : ((NamespacedEntityId) entityId).getNamespaceId();
 
       // validate the profile and publish the necessary metadata change if the profile exists in the property
       Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(namespaceId, propertyMap);
-      // get the old profile information from the previous properties
-      Optional<ProfileId> oldProfile = SystemArguments.getProfileIdFromArgs(namespaceId, oldProperties);
       if (profile.isPresent()) {
         ProfileId profileId = profile.get();
+        // if it is instance level set, the profile has to be in SYSTEM scope, so throw BadRequestException when
+        // setting a USER scoped profile
+        if (isInstanceLevel && !propertyMap.get(SystemArguments.PROFILE_NAME).startsWith("SYSTEM")) {
+          throw new BadRequestException(String.format("Cannot set user scope profile at instance level, profile " +
+                                                        "property must look like SYSTEM:[profile-name], now found: %s",
+                                                      propertyMap.get(SystemArguments.PROFILE_NAME)));
```

Comment:
you can use ProfileId's new getScopedName() method

Reply:
As mentioned above, the profileId's scopedName will be at SYSTEM scope, since for instance level we construct the profileid using NamespaceId.SYSTEM directly, because we do not know the namespace id for the profile

Reply:
As mentioned, the getScopedName() for instance level profile will always be Namespace.SYSTEM since we pass it as the namespace to SystemArguments.